integration_tests/steps/two_sided_market.py

```diff
@@ -134,20 +140,19 @@ def energy_rate_average_between_min_and_max_ess_pv(context):
     house2 = next(filter(lambda x: x.name == "House 2", context.simulation.area.children))
     pv = next(filter(lambda x: "H2 PV" in x.name, house2.children))
 
-    storage_rates_set = set()
-    pv_rates_set = set()
     for market in house1.past_markets:
         for trade in market.trades:
             if trade.buyer == storage.name:
-                storage_rates_set.add(trade.offer.price / trade.offer.energy)
+                initial = storage.strategy.bid_update.initial_rate[market.time_slot]
+                final = storage.strategy.bid_update.final_rate[market.time_slot]
+                assert initial <= (trade.offer.price / trade.offer.energy) <= final
 
     for market in house2.past_markets:
         for trade in market.trades:
             if trade.seller == pv.name:
-                pv_rates_set.add(trade.offer.price / trade.offer.energy)
-
-    assert all([RATE_LOWER_THRESHOLD < rate < RATE_UPPER_THRESHOLD for rate in storage_rates_set])
-    assert all([RATE_LOWER_THRESHOLD < rate < RATE_UPPER_THRESHOLD for rate in pv_rates_set])
+                initial = storage.strategy.bid_update.initial_rate[market.time_slot]
+                final = storage.strategy.bid_update.final_rate[market.time_slot]
+                assert initial <= (trade.offer.price / trade.offer.energy) <= final
```

Comment:
IMO this is a too unconstrained assertion. Yes, it should be in between initial and final, but arent there rates that are in a certain range? 
Of course we will probably change this behaviour when we implement the changes tot the 2 sided market behaviour. But still, should be the trade rate be somewhere around 15?

Reply:
Actually, mean mean the middle here, so initially, the rates were in the middle between initial and final. I remember when I first changed this, was the trade rate 15 because the initial was 30 and the final rate was 15. 
Now you just check whether the rate is between the two, so it could be 1 or also 29. 
Maybe the behaviour of the trade rate changed and the trade rate is fluctuating a lot, but then, please rename the step function. Thanks!

Reply:
EDIT: There is an error in my comment. This is the correct one:
Actually, mean mean the middle here, so initially, the rates were in the middle between initial and final. I remember when I first changed this, the trade rate was 15 because the initial was 30 and the final rate was 0.
Now you just check whether the rate is between the two, so it could be 1 or also 29.
Maybe the behaviour of the trade rate changed and the trade rate is fluctuating a lot, but then, please rename the step function. Thanks!

Reply:
May be I have misunderstood it in place but the scenario is testing two sided market for one ESS and one PV `Scenario: One storage, one pv`. It was previously also checking if ESS not violating its buying THRESHOLD.
What I tried to do here is to rather than comparing with hard coded THRESHOLD, Im now  comparing with threshold(i.e. initial and final buying rate) that is set at the time of initialising ESS strategy.

I hope I didn't deviated the initial target of this integration test

Reply:
Well, you kind of did. 
The step function is called: "the energy rate for all the trades is the mean of max and min pv/storage rate" 
As you said, now it is checking, whether it is not violating the initial-final limit. These two are different. 
So, either you rename the function OR you look in which range the trade_rates are and readjust the RATE_LOWER_THRESHOLD and RATE_UPPER_THRESHOLD .

Reply:
Sorry, I couldn't find any `mean` in the diff. It is accumulating the rate at which pv_traded_energy and ess_traded_energy. And the asserting if all of these trades are in between hard-coded RATE_LOWER_THRESHOLD & RATE_UPPER_THRESHOLD.

BTW: its been renamed to `the energy rate for all trades are in between initial and final buying rate of storage`